### PR TITLE
ref(performance): update api for page error alerts

### PR DESCRIPTION
**Pull request by DominikB2014**
Would like some thoughts on this.
This PR 
1. Changes the page error context to be more generic and use pagealerts
2. Updates Api to allow page alerts to be of any type, you can specify type:'error', type:'message', etc

In the future we will
1. Adds an optional param to allow alerts to be dismissed 
which will be used in resource module
image
2. Support multiple alerts at once

Some questions:
- Do we generally agree with this change?
- Should we default to type:'error' as this seems to be by far the most common option atm
- The dismissing functionality resets when you re-load the context, for example switching from resource sidebar item -> another sidebar item -> back to resource sidebar item. SHould we store the dismissed alerts somewhere more globally to avoid this?

### CODE REVIEW

File: static/app/utils/performance/contexts/pageError.spec.tsx
Comment (gggritso) [inline static/app/utils/performance/contexts/pageError.spec.tsx:1]: The file name doesn't match the component anymore

File: static/app/utils/performance/contexts/pageError.tsx
Comment (gggritso) [inline static/app/utils/performance/contexts/pageError.tsx:1]: The file name doesn't match the component anymore

File: static/app/utils/performance/contexts/pageError.tsx [hunk 1]
```diff
@@ -1,49 +1,78 @@
...
+import {createContext, Fragment, useContext, useRef, useState} from 'react';
 import {Theme} from '@emotion/react';
 
 import {Alert} from 'sentry/components/alert';
+import {IconClose} from 'sentry/icons';
 
-export type PageAlert = {
-  message: string;
+export type PageAlertOptions = {
+  message: string | undefined;
```
Comment (gggritso) [inline static/app/utils/performance/contexts/pageError.tsx:8]: 1. Why `| undefined`? Is that the mechanism for removing an alert?
2. I think it should be `React.ReactNode` or something similar, so hook users can pass in formatted DOM elements

File: static/app/utils/performance/contexts/pageError.tsx [hunk 2]
```diff
@@ -1,49 +1,78 @@
...
   type: keyof Theme['alert'];
+  dismissible?: boolean;
```
Comment (gggritso) [inline static/app/utils/performance/contexts/pageError.tsx:10]: Looks like other parts of the code use `isDismissible` instead

**Review by gggritso — COMMENTED** [review]
Makes sense to me, I like it! I think it's a good idea to have _one place_ at the top of the page to aggregate and show alert banners. `PageAlert` makes sense to me.

I think it's fine to have `"error"` be the default level, but I think an _even cleaner_ solution would be to export multiple setters from the hook: `setError`, setWarning`, `setInfo`. A reasonable API would be `setPageError(message: React.ReactNode, options: {isDismissable?: boolean)`. This would shrink the PR a bit, too

In order for `dismissable` to be useful, it needs to persist between reloads. IMO you can remove that functionality from this PR until there's a solution for that, since it's net new functionality

Lastly, if I'm reading this right, it only shows one banner at a time, which is limiting! Even more power would be in consolidating, sorting, and showing multiple alerts. Maybe a good follow-up?

File: static/app/utils/performance/contexts/pageError.spec.tsx
Comment (DominikB2014) [reply static/app/utils/performance/contexts/pageError.spec.tsx:1]: 👍 good catch will update that!

File: static/app/utils/performance/contexts/pageError.tsx (hunk 1 above)
Comment (DominikB2014) [reply static/app/utils/performance/contexts/pageError.tsx:8]: 1. Yeah pretty much, and also to support things like without further refactor https://github.com/getsentry/sentry/pull/61881/files#diff-406fb42bb51b542389d6d77c4d299d2c61606d9773204c1d4cc9689aac402188R70
where msg is `string | undefined`. 

2. `React.ReactNode` is a great idea, i'll make the change!

File: static/app/utils/performance/contexts/pageError.tsx (hunk 2 above)
Comment (DominikB2014) [reply static/app/utils/performance/contexts/pageError.tsx:10]: 👍

**Comment by DominikB2014** [comment]
> Makes sense to me, I like it! I think it's a good idea to have _one place_ at the top of the page to aggregate and show alert banners. `PageAlert` makes sense to me.
> 
> I think it's fine to have `"error"` be the default level, but I think an _even cleaner_ solution would be to export multiple setters from the hook: `setError`, setWarning`, `setInfo`. A reasonable API would be `setPageError(message: React.ReactNode, options: {isDismissable?: boolean)`. This would shrink the PR a bit, too
> 
100%, i'll make that change

> In order for `dismissable` to be useful, it needs to persist between reloads. IMO you can remove that functionality from this PR until there's a solution for that, since it's net new functionality

Yeah I was on the fence about that, this can probably be easily done via localstorage or something, I agree and i'll move this to another PR

> 
> Lastly, if I'm reading this right, it only shows one banner at a time, which is limiting! Even more power would be in consolidating, sorting, and showing multiple alerts. Maybe a good follow-up?
Yup 100%, great follow up, should be an easy add.

**Review by DominikB2014 — COMMENTED** [review]
@gggritso I'll also

**Comment by DominikB2014** [comment]
@gggritso I think I addressed most of your comments, dismissible alerts and multiple alert support can be a future improvement!

File: static/app/utils/performance/contexts/pageAlert.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    setPageAlert({message, type: 'success'});
+  };
+
+  const setPageWarning = (message: string | undefined) => {
+    setPageAlert({message, type: 'warning'});
+  };
+
+  const setPageError = (message: string | undefined) => {
+    setPageAlert({message, type: 'error'});
+  };
```
Comment (gggritso) [inline static/app/utils/performance/contexts/pageAlert.tsx:47]: You could clean this repetition up by iterating through the types, and creating functions on the fly!

File: static/app/utils/performance/contexts/pageAlert.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  setPageWarning: (message: string | undefined) => void;
+  pageAlert?: PageAlertOptions;
+}>({
+  pageAlert: undefined,
+  setPageError: (_: string | undefined) => {},
+  setPageInfo: (_: string | undefined) => {},
+  setPageMuted: (_: string | undefined) => {},
+  setPageSuccess: (_: string | undefined) => {},
+  setPageWarning: (_: string | undefined) => {},
+});
```
Comment (gggritso) [inline static/app/utils/performance/contexts/pageAlert.tsx:25]: The types here are a little funky! At a minimum, is it possible to extract `(_: string | undefined) => {}` and `(message: string | undefined) => void` into a type?

Comment (gggritso) [reply static/app/utils/performance/contexts/pageAlert.tsx:25]: Also, these should accept a `React.ReactNode` to match the component, no?

**Review by gggritso — COMMENTED** [review]
Looking good, just a few more nits!

File: static/app/utils/performance/contexts/pageAlert.tsx (hunk 3 above)
Comment (DominikB2014) [reply static/app/utils/performance/contexts/pageAlert.tsx]: Hmm, I really like the looping idea, but I don't think there's a way to loop through a union type in Typescript.
Perhaps I can just declare a const, and just loop through that?

Comment (gggritso) [reply static/app/utils/performance/contexts/pageAlert.tsx]: Sorry for the delay! You can loop through union types in TypeScript! It's a little awkward [but possible](https://www.typescriptlang.org/play?target=6#code/MYewdgzgLgBAMgUQGoLgZRgXhgbQOQCWYAZiHjAIYQyiRQA0MeA7hQE5hEDm5VN40RnggBXYMACmECL2q1BTCWzYg2s-nQC6lOQKgBuAFCGoATwAOE+BIBuEgDZYYZyyGLxkqNDjAiAtgBGSppGJhZWcLYOAMIU9vYBFMAA1tTYAN6GMNm4kXb2AHIUflZE1vk6MAAGEBJQAAoUXBIAJOmx5gRQcQQAXhIAPHkORSUAfAC+VZoAXDAAFCXSTRIA-HPQbNwAlFhjMDYgBAAmRhPG8rDAcQlJqU4A8gEAVhLAUAB0xCp+CGBQWyk80QKHQHz8FHM83sUUKxSsmH2mRyGmgNBuiRSowR1VqDRWbWunW69j6EmhsOx2ymlUuRhRl3R8UxySciykEBWGwBOz2C3SMCWnOajBcEjmMPy2Im21CKLYdREHFw12Zd2xjFVtxSIUMMt2fBEYGSYBAzDAlWG9liapSEFChmIRveBHA6OJPX68023K2YC422ROQVUCVFs2H2AAAt2ABBKDzAAM2w+UBAAFVzJY2LFavNdgBqGAR0QBTbceYARlleouAhAMI+9hAXHmWpZEA+eMazQA6uxOP75ngow5m3htrKgA). That's my first stab at it, just to give an idea, maybe a cleaner version is possible

File: static/app/utils/performance/contexts/pageAlert.tsx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  setPageSuccess: PageAlertSetter;
+  setPageWarning: PageAlertSetter;
+  pageAlert?: PageAlertOptions;
+}>({
+  pageAlert: undefined,
+  setPageError: (_: React.ReactNode | undefined) => {},
+  setPageInfo: (_: React.ReactNode | undefined) => {},
+  setPageMuted: (_: React.ReactNode | undefined) => {},
+  setPageSuccess: (_: React.ReactNode | undefined) => {},
+  setPageWarning: (_: React.ReactNode | undefined) => {},
```
Comment (gggritso) [inline static/app/utils/performance/contexts/pageAlert.tsx:28]: Should this be `void` instead? The return value is not usable, is it?

```suggestion
  setPageError: (_: React.ReactNode | undefined) => void,
  setPageInfo: (_: React.ReactNode | undefined) => void,
  setPageMuted: (_: React.ReactNode | undefined) => void,
  setPageSuccess: (_: React.ReactNode | undefined) => void,
  setPageWarning: (_: React.ReactNode | undefined) => void
```

File: static/app/utils/performance/contexts/pageAlert.tsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      {children}
+    </pageErrorContext.Provider>
+  );
+}
+
+export function PageAlert() {
+  const {pageAlert: pageError} = useContext(pageErrorContext);
+  if (!pageError || !pageError.message) {
+    return null;
+  }
```
Comment (gggritso) [inline static/app/utils/performance/contexts/pageAlert.tsx:73]: Why the rename to `pageError`?

**Review by gggritso — APPROVED** [review]
Sorry for the delay on this, very nice! Nits only, nothing serious 👍🏻 Great work

File: static/app/utils/performance/contexts/pageAlert.tsx (hunk 6 above)
Comment (DominikB2014) [reply static/app/utils/performance/contexts/pageAlert.tsx]: This was just the result of vscode doing an automatic variable rename, updated! nice catch!

**Pull request merged**